algoliasearch/src/main/java/com/algolia/search/saas/Client.java

```diff
@@ -111,18 +113,26 @@ public Client(@NonNull String applicationID, @NonNull String apiKey, String[] ho
             setReadHosts(hosts);
             setWriteHosts(hosts);
         } else {
-            setReadHosts(
-                    applicationID + "-dsn.algolia.net",
-                    applicationID + "-1.algolianet.com",
-                    applicationID + "-2.algolianet.com",
-                    applicationID + "-3.algolianet.com"
-            );
-            setWriteHosts(
-                    applicationID + ".algolia.net",
+            // Initialize hosts to their default values.
+            //
+            // NOTE: The host list comes in two parts:
+            //
+            // 1. The fault-tolerant, load-balanced DNS host.
+            // 2. The non-fault-tolerant hosts. Those hosts must be randomized to ensure proper load balancing in case
+            //    of the first host's failure.
+            //
+            List<String> fallbackHosts = Arrays.asList(
                     applicationID + "-1.algolianet.com",
                     applicationID + "-2.algolianet.com",
                     applicationID + "-3.algolianet.com"
             );
+            Collections.shuffle(fallbackHosts);
+            readHosts = new ArrayList<>(fallbackHosts.size() + 1);
+            readHosts.add(applicationID + "-dsn.algolia.net");
+            readHosts.addAll(fallbackHosts);
+            writeHosts = new ArrayList<>(fallbackHosts.size() + 1);
+            writeHosts.add(applicationID + ".algolia.net");
```

Comment:
`-dsn` is only for the search.
write must be made to the cluster not the slaves